File: 347_top_k_frequent_elements/top_k_frequent_elements.py
from typing import List, Dict


class Solution:
    def top_k_frequent_elements(self, nums: List[int], k: int) -> List[int]:
        freq = [[] for i in range(len(nums) + 1)]
        
        count: Dict[int:int] = {}
        for n in nums:
            count[n] = count.get(n, 0) + 1

        for n, c in count.items():
            freq[c].append(n)
        
        res = []
        for i in range(len(freq)-1, 0, 1):
            for n in freq[i]:
                res.append(n)
                if len(res) == k:
                    return res


if __name__ == "__main__":
    solution = Solution()
    nums, k = [1,1,1,2,2,3], 2
    solution.top_k_frequent_elements(nums, k)

File: 347_top_k_frequent_elements/top_k_frequent_elements.go
package main

import "fmt"

type Solution struct{}

func (s Solution) topKFrequentElements(nums []int, k int) []int {

	count := make(map[int]int)
	for _, n := range nums {
		count[n]++
	}

	freq := make([][]int, len(nums)+1)
	for n, c := range count {
		freq[c] = append(freq[c], n)
	}

	res := []int{}
	for i := len(freq) - 1; i >= 0; i-- {
		for _, n := range freq[i] {
			res = append(res, n)
			if len(res) == k {
				return res
			}
		}
	}
	return res
}

func main() {
	s := Solution{}
	nums := []int{1, 1, 1, 2, 2, 3}
	k := 2
	fmt.Println(s.topKFrequentElements(nums, k))
}
